src/awkward/_typetracer.py

```diff
@@ -198,6 +201,104 @@ def __str__(self):
+            # touching data implies that the shape will be touched as well
+            # implemented here so that the codebase doesn't need to be filled
+            # with calls to both methods everywhere
+            self._shape_touched_set.add(label)
+            self._shape_touched.append(label)
+            self._data_touched_set.add(label)
+            self._data_touched.append(label)
+
+
+def _attach_report(layout, form, report):
+    if isinstance(layout, (ak.contents.BitMaskedArray, ak.contents.ByteMaskedArray)):
+        assert isinstance(form, (ak.forms.BitMaskedForm, ak.forms.ByteMaskedForm))
+        layout.mask.data.form_key = form.form_key
+        layout.mask.data.report = report
+        _attach_report(layout.content, form.content, report)
+
+    elif isinstance(layout, ak.contents.EmptyArray):
+        assert isinstance(form, ak.forms.EmptyForm)
+        layout.mask.data.form_key = form.form_key
+        layout.mask.data.report = report
+
+    elif isinstance(layout, (ak.contents.IndexedArray, ak.contents.IndexedOptionArray)):
+        assert isinstance(form, (ak.forms.IndexedForm, ak.forms.IndexedOptionForm))
+        layout.index.data.form_key = form.form_key
+        layout.index.data.report = report
+        _attach_report(layout.content, form.content, report)
+
+    elif isinstance(layout, ak.contents.ListArray):
+        assert isinstance(form, ak.forms.ListForm)
+        layout.starts.data.form_key = form.form_key
+        layout.starts.data.report = report
+        layout.stops.data.form_key = form.form_key
+        layout.stops.data.report = report
+        _attach_report(layout.content, form.content, report)
+
+    elif isinstance(layout, ak.contents.ListOffsetArray):
+        assert isinstance(form, ak.forms.ListOffsetForm)
+        layout.offsets.data.form_key = form.form_key
+        layout.offsets.data.report = report
+        _attach_report(layout.content, form.content, report)
+
+    elif isinstance(layout, ak.contents.NumpyArray):
+        assert isinstance(form, ak.forms.NumpyForm)
+        layout.data.form_key = form.form_key
+        layout.data.report = report
+
+    elif isinstance(layout, ak.contents.RecordArray):
+        assert isinstance(form, ak.forms.RecordForm)
+        for x, y in zip(layout.contents, form.contents):
+            _attach_report(x, y, report)
+
+    elif isinstance(layout, (ak.contents.RegularArray, ak.contents.UnmaskedForm)):
+        assert isinstance(form, (ak.forms.RegularForm, ak.forms.UnmaskedForm))
+        _attach_report(layout.content, form.content, report)
+
+    elif isinstance(layout, ak.contents.UnionArray):
+        assert isinstance(form, ak.forms.UnionForm)
+        layout.tags.data.form_key = form.form_key
+        layout.tags.data.report = report
+        layout.index.data.form_key = form.form_key
+        layout.index.data.report = report
+        for x, y in zip(layout.contents, form.contents):
+            _attach_report(x, y, report)
+
```

Comment:
Right! But this kind of error (our programming error, rather than a user's/downstream developer's usage error) is `AssertionError`. I implemented this in d66e7ccf6f792d0316d2c243e69704778c633dce.